### PR TITLE
[Fix] #353: Evaluate Buildscan Directly When Root is Evaluated

### PR DESCRIPTION
**Pull request by bmarcaur**
## Before this PR
Issue is described at length: https://github.com/palantir/gradle-git-version/issues/353

## After this PR
Usages of `git-version` on submodules no longer errors out.

## Possible downsides?
It's possible that this [check](https://github.com/palantir/gradle-git-version/blob/develop/src/main/java/com/palantir/gradle/gitversion/GitVersionRootPlugin.java#L29) was meant to fail when a user does not apply the plugin at the root, but as it stands all it does is runtime check that our code properly calls out to a project's root and then applies it. So an alternative fix is to hard error when this is called in submodules. However, fixing that "bug" is a breaking change.

### CODE REVIEW

**Comment by bmarcaur** [comment]
Note for the reviewers: I wrote the test, observed the failing behavior, and then I applied the fix. So this definitely exhibits and fixes the behavior described in #353

File: changelog/@unreleased/pr-541.v2.yml [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+    [Fix] 353: Evaluate Buildscan Directly When Root is Evaluated
+
+    ## Before this PR
+    Issue is described at length: https://github.com/palantir/gradle-git-version/issues/353
+
+    ## After this PR
+    Usages of `git-version` on submodules no longer errors out.
+
+    ## Possible downsides?
+    It's possible that this [check](https://github.com/palantir/gradle-git-version/blob/develop/src/main/java/com/palantir/gradle/gitversion/GitVersionRootPlugin.java#L29) was meant to fail when a user does not apply the plugin at the root, but as it stands all it does is runtime check that our code properly calls out to a project's root and then applies it. So an alternative fix is to hard error when this is called in submodules. However, fixing that "bug" is a breaking change.
```
Comment (carterkozak) [inline changelog/@unreleased/pr-541.v2.yml]:
```suggestion
    Evaluate Buildscan directly when the root project is evaluated
```

File: src/groovy/groovy/com/palantir/gradle/gitversion/BuildScanPluginInterop.groovy [hunk 2]
```diff
@@ -21,24 +21,36 @@ import org.gradle.api.Project
...
 class BuildScanPluginInterop {
     static void addBuildScanCustomValues(Project rootProject, Supplier<Map<String, String>> customValues) {
-        // After evaluate because while we can detect the <5.x com.gradle.build-scan project on the root project,
-        // there is no way to detect the >6.x com.gradle.enterprise settings plugins using withPlugin
-        rootProject.afterEvaluate {
-            def gradleEnterpriseExtension = rootProject.extensions.findByName('gradleEnterprise')
+        // Fix #353: Detect when the root project is already evaluated, thus anything is afterEvaluate
+        // and we can just execute immediately. This is due to afterEvaluate changes in Gradle 7.
+        //
+        // Note: We cannot use hasCompleted() because we need to support gradle 5
```
Comment (carterkozak) [inline src/groovy/groovy/com/palantir/gradle/gitversion/BuildScanPluginInterop.groovy]: > we need to support gradle 5

Why?

Comment (bmarcaur) [reply src/groovy/groovy/com/palantir/gradle/gitversion/BuildScanPluginInterop.groovy]: For no other reason that I did not want to take away behavior as I suspected that would require a major version bump. The previous comments seemed to imply that having this work on Gradle 5 was something to be desired? IDC TBH

Comment (bmarcaur) [reply src/groovy/groovy/com/palantir/gradle/gitversion/BuildScanPluginInterop.groovy]: I clarified the language, not sure that the upside is worth the loss of gradle 5 at this point.

**Review by carterkozak — APPROVED** [review]

**Pull request merged**